File: go/116-secret-handshake/secret_handshake.go
package secret

import (
	"fmt"
	"slices"
)

func Handshake(code uint) []string {
	num := code
	var binary string
	for num > 0 {
		remainder := num % 2
		binary = fmt.Sprintf("%d%s", remainder, binary)
		num = num / 2
	}

	// binary := fmt.Sprintf("%b", code)

	var result []string
	for i := len(binary) - 1; i >= 0; i-- {
		if binary[i] == '1' {
			switch len(binary) - 1 - i {
			case 0:
				result = append(result, "wink")
			case 1:
				result = append(result, "double blink")
			case 2:
				result = append(result, "close your eyes")
			case 3:
				result = append(result, "jump")
			case 4:
				slices.Reverse(result)
				return result
			}
		}
	}

	return result
}

File: ruby/077-secret-handshake/secret_handshake.rb
class SecretHandshake
  # @param[Numeric]num
  def initialize(num)
    @num = num
  end

  def commands
    return [] if @num.to_s.to_i != @num

    actions = []
    binary = []
    n = 16
    while n > 0
      div = @num / n
      binary << div
      @num -= n if div == 1
      n = n / 2
    end

    binary.reverse.each_with_index do |bin, index|
      next if bin != 1

      case index
      when 0 then actions << "wink"
      when 1 then actions << "double blink"
      when 2 then actions << "close your eyes"
      when 3 then actions << "jump"
      when 4 then actions.reverse!
      end
    end

    actions
  end
end